### ISSUE TITLE
Dynamic Queries in dataloader

### ISSUE THREAD

**Issue by tvvignesh** [facebook/dataloader#137]
As dataloader uses just the key to identify whether to load from cache or the db, how do we handle dynamic queries?

An example use case:
1) Through graphQL, I apply filters to get only certain fields

2) On the server side, instead of doing a static SQL query, I form the queries dynamically based on the graphql input (Using sequelize for doing this). So, the number of fields returned is not always the same for the same db function (with same key). For eg, for the same employee code, I can ask 3 fields in db or 5 fields and queries are formed dynamically.

3) Lets assume I increase the fields requested now from 3 to 5 in the input, and nothing changes, the dataloader finds the key and loads it from existing cache. How will I handle the 2 missing columns.

To summarize, how will I handle different number of columns for same key in dataloader?

I can do SELECT * to get all fields and then filter in the code but just wanted to know if there is anyway in filtering via query itself.

**Comment by tvvignesh** [comment]
Never mind, I guess since dataloader's lifetime is per request, this must not be an issue I guess.

**Comment by timscott** [comment]
@tvvignesh How are you solving this? I have the same problem. I want to build my select clause from `info.fieldNodes` passed into the resolver. How do I get it into the body of the DataLoader callback? Or is there some other approach that you're using. Thanks!

**Comment by ChrisLahaye** [comment]
@timscott did you find a solution for this?

**Comment by tvvignesh** [comment]
@timscott, @ChrisLahaye: Sorry for the late reply. I missed your message. The dynamic queries don't matter much since the lifetime of a dataloader is meant to be just one request and is not meant to be used across requests. 

To summarize, there is no solution to this and no solution is needed unless your queries are heavily changing in just a single request-response cycle. Hope this clarifies.

**Comment by timscott** [comment]
@tvvignesh @ChrisLahaye: Actually, the dynamic queries can matter even in the context of one request. Consider the following:

```graphql
query {
  thing {
    personA {
      email
    }
    persionB {
      firstName
    }
  }
}
```

If you build a query on the person table dynamically from requested fields, then if personA and personB are the same person, and you use only the ID as the cache key, then you might get an error like this:

```Cannot return null for non-nullable field Person.firstName.```

I have solved it like so:

```js
const key = {id, args, selectedFields};
return this.dataLoader.load(key);
```
And then provide a custom ```cacheKeyFn```. There is a discussion about how to do that at https://github.com/graphql/dataloader/issues/75.
